### PR TITLE
Fixing issue with explode_outer position not nulling position entries of null rows

### LINKED ISSUE #7721: [BUG] explode_outer_position doesn't match to Spark's counterpart
**Describe the bug**
In cuDF,  `explode_outer_position` will mark the position values of empty rows with `0`. Meanwhile, the position values of empty rows are marked as `null` in Spark.

**Steps/Code to reproduce bug**
For input data like:
 * [[5,null,15], 100]
 * [null,        200]
 * [[],          300]
 
cuDF returns
 * [0,     5,    100]
 * [1,  null,    100]
 * [2,    15,    100]
 * [0,  null,    200]
 * [0,  null,    300]

But Spark returns
 * [0,     5,    100]
 * [1,  null,    100]
 * [2,    15,    100]
 * [null,  null,    200]
 * [null,  null,    300]

### PR DESCRIPTION
**Pull request by hyperbolic2346**
`explode_outer` supports writing a position column, but if the row was null it would incorrectly set the position to 0 and the row valid. Instead, it should null that position row as well. Luckily the null column matches 100% with the null column of the exploded column, so we can just copy it after it is created.

Fixes #7721

### CODE REVIEW

**Comment by sperlingxx** [comment]
This change looks good in terms of spark-rapids.

**Comment by jrhemstad** [comment]
@kkraus14 is there a Pandas equivalent for "explode_outer" that we should be concerned with matching its semantics?

**Comment by jrhemstad** [comment]
So `cudf::explode_outer` -> `pd.explode`, `cudf::explode` -> `pd.?`

**Comment by kkraus14** [comment]
> `cudf::explode` -> `pd.?`

We don't have a need for an equivalent of a non-outer explode at this time.

**Comment by jrhemstad** [comment]
What's more, @hyperbolic2346's original implementation would have resulted in:

>       a  b
> 0     1  1
> 0     2  1
> 0     3  1
> 1     4  2
> 1     5  2
> 0   NaN  3
> 0  <NA>  4

i.e., zeros in the empty/null list positions. So that didn't do what Pandas wanted either.

**Comment by jrhemstad** [comment]
In fact, @kkraus14 the example you gave is totally different. 


For input data like:
 * [[5,null,15], 100]
 * [null,        200]
 * [[],          300]
 
@hyperbolic2346 's original implementation returns
 * [0,     5,    100]
 * [1,  null,    100]
 * [2,    15,    100]
 * [0,  null,    200]
 * [0,  null,    300]
 
 The new one would do:
  * [0,     5,    100]
 * [1,  null,    100]
 * [2,    15,    100]
 * [null,  null,    200]
 * [null,  null,    300]

**Comment by jrhemstad** [comment]
Ah, perhaps Pandas doesn't need the `cudf::explode_outer_position`. You could just do `explode_outer` and prepend the index as an additional column that would yield the result from @kkraus14 's example.

**Comment by kkraus14** [comment]
> Ah, perhaps Pandas doesn't need the `cudf::explode_outer_position`. You could just do `explode_outer` and prepend the index as an additional column that would yield the result from @kkraus14 's example.

Yes, we do not need position handling, just `explode_outer` and we'll pass the index column(s) as needed.

File: cpp/src/lists/explode.cu [hunk 1]
```diff
@@ -62,22 +62,26 @@ std::unique_ptr<table> build_table(
...
-    columns.insert(columns.begin() + explode_column_idx,
-                   std::make_unique<column>(
-                     data_type(type_to_id<size_type>()), position_size, position_array->release()));
+    // the null mask for position matches the exploded column's gather map, so copy it over
+    rmm::device_buffer nullmask =
+      explode_col_gather_map ? copy_bitmask(*inserted->get()) : rmm::device_buffer(0, stream);
+    columns.insert(
+      columns.begin() + explode_column_idx,
+      std::make_unique<column>(
+        data_type(type_to_id<size_type>()), position_size, position_array->release(), nullmask));
```
Comment (jrhemstad) [inline cpp/src/lists/explode.cu:84]:
```suggestion
        data_type(type_to_id<size_type>()), position_size, position_array->release(), std::move(nullmask)));
```

**Review by jrhemstad — APPROVED** [review]

**Comment by hyperbolic2346** [comment]
@gpucibot merge

**Pull request merged**